### ISSUE TITLE
Dialog windows are not modal

### ISSUE THREAD

**Issue by LordOfDragons** [beeware/toga#2446]
### Describe the bug

Create a dialog like this:
```
async def on_button_test(widget):
    await self.main_window.info_dialog(title="Dialog", message="An Info Dialog")
```
You can now click on the main window behind the dialog  and the main window is activated although the dialog is still floating above the window. This allows the dialog to be shown again or other actions to be triggered although the user should be blocked in the dialog.

Either change the behavior of dialogs to be modal or add a "modal={True|False}" property to "info_dialog" (and all similar functions) to enforce this.

### Steps to reproduce

1) Create self.main_window.info_dialog
2) Click main window
3) Main window is activated and can be interacted with. 

### Expected behavior

Main window can not be activated (dialog keeps focus)

### Environment

Linux (GenToo). Toga in virtual environment.

**Comment by freakboy3742** [comment]
Thanks for the report. Looks like this might have been an oversight when we modified the dialog implementation to make it testable. 

It looks like adding `self.native.set_modal(True)` around L33 of `gtk/src/toga_gtk/dialogs.py` will make the dialog modal without breaking the testability of the dialog. There might be a similar change needed for the non-message dialogs (e.g. stack trace and file selection dialogs).

**Comment by LordOfDragons** [comment]
I also just noticed the close button in the default About dialog has no effect (aka not closing the dialog).

**Comment by freakboy3742** [comment]
I've noticed that as well, but not logged it. I've fixed *that* problem as part of #2244 - the specific change is [this one](https://github.com/beeware/toga/pull/2244/files#diff-49b823fae670b3d17a24b95c5e3037455a0a21cfa2998d6aa031b6003f1bd671R317) - however, that PR is taking a lot longer to land than originally planned. If that change were to slip into a fix for this issue, that would probably be for the better.
